File: src/Candidate.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.*;

enum Gender {
    MASCULINE,
    FEMININE;
}

enum CandidateSituation {
    ELECTED,
    NOT_ELECTED,
    INVALID;
}

public class Candidate {
    private int cityCode;
    private int jobCode;
    private int candidateNumber;
    private String candidateName;
    private int partyNumber;
    private String partyAcronym;
    private int federationNumber;
    private LocalDate birthDate;
    private CandidateSituation candidateSituation;
    private Gender gender;
    private int candidateVotes;
    private long candidateAge;

    public Candidate(String[] values) {
        this.cityCode = Integer.parseInt(values[0]);
        this.jobCode = Integer.parseInt(values[1]);
        this.candidateNumber = Integer.parseInt(values[2]);
        this.candidateName = values[3];
        this.partyNumber = Integer.parseInt(values[4]);
        this.partyAcronym = values[5];
        this.federationNumber = Integer.parseInt(values[6]);

        String dateStr = values[7];

        // convert string to LocalDate
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
        this.birthDate = LocalDate.parse(dateStr, formatter);

        if (values[8].equals("2") || values[8].equals("3")) {
            this.candidateSituation = CandidateSituation.ELECTED;
        } else if (values[8].equals("-1")) {
            this.candidateSituation = CandidateSituation.INVALID;
        } else {
            this.candidateSituation = CandidateSituation.NOT_ELECTED;
        }

        if (values[9].equals("2")) {
            this.gender = Gender.MASCULINE;
        } else if (values[9].equals("4")) {
            this.gender = Gender.FEMININE;
        }
    }

    /* 
     * create candidates and save them in each political party
     */
    public static HashMap<Integer, Candidate> CreateCandidates(List<String[]> values, HashMap<Integer, PoliticalParty> politicalParties) {
        HashMap<Integer, Candidate> candidates = new HashMap<Integer, Candidate>();

        for (String[] value : values) {
            Candidate candidate = new Candidate(value);
            candidates.put(candidate.candidateNumber, candidate);

            PoliticalParty candidateParty = politicalParties.get(candidate.partyNumber);
            candidateParty.addCandidate(candidate);
        }

        return new HashMap<>(candidates);
    }

    // getters
    public int getCityCode() {
        return cityCode;
    }

    public int getJobCode() {
        return jobCode;
    }

    public int getCandidateNumber() {
        return candidateNumber;
    }

    public String getCandidateName() {
        return candidateName;
    }

    public int getPartyNumber() {
        return partyNumber;
    }

    public String getPartyAcronym() {
        return partyAcronym;
    }

    public int getFederationNumber() {
        return federationNumber;
    }

    public LocalDate getBirthDate() {
        return birthDate;
    }

    public CandidateSituation getCandidateSituation() {
        return candidateSituation;
    }

    public Gender getGender() {
        return gender;
    }

    public int getCandidateVotes() {
        return candidateVotes;
    }
    // end getters

    public void incCandidateVotes(int votes) {
        candidateVotes += votes;
    }

    public void setCandidateAge(LocalDate electionDate) {
        candidateAge = ChronoUnit.YEARS.between(birthDate, electionDate);
    }

    public long getCandidateAge() {
        return candidateAge;
    }
}

File: src/candidate.cpp
#include "candidate.hpp"

map<int, Candidate> CreateCandidates(
    const vector<vector<string> > &values,
    const map<int, PoliticalParty> &politicalParties)
    {

    }

int Candidate::getCityCode() {
    return cityCode;
}

int Candidate::getJobCode() {
    return jobCode;
}

int Candidate::getCandidateNumber() {
    return candidateNumber;
}

string Candidate::getCandidateName() {
    return candidateName;
}

int Candidate::getPartyNumber() {
    return partyNumber;
}

string Candidate::getPartyAcronym() {
    return partyAcronym;
}

int Candidate::getFederationNumber() {
    return federationNumber;
}

tm Candidate::getBirthDate() {
    return birthDate;
}

CandidateSituation Candidate::getCandidateSituation() {
    return candidateSituation;
}

Gender Candidate::getGender() {
    return gender;
}

int Candidate::getCandidateVotes() {
    return candidateVotes;
}

void Candidate::incCandidateVotes(const int &votes) {
    candidateVotes += votes;
}

void Candidate::setCandidateAge(const tm &electionDate) {
    int age = electionDate.tm_year - birthDate.tm_year;

    // if the birthday hasnt happened in the year yet
    if (electionDate.tm_mon < birthDate.tm_mon ||
        (electionDate.tm_mon == birthDate.tm_mon && electionDate.tm_mday < birthDate.tm_mday)) {
        age--;
    }
    candidateAge = age;
}

long Candidate::getCandidateAge() {
    return candidateAge;
}

int Candidate::getMostVotedPosition() {
    return mostVotedPosition;
}

void Candidate::setMostVotedPosition(const int &mostVotedPosition) {
    this->mostVotedPosition = mostVotedPosition;
}

// muito feio!!
tm parseDate(const string& dateStr) {
    tm date = {};
    istringstream ss(dateStr);
    char sep1, sep2;

    ss >> date.tm_mday >> sep1 >> date.tm_mon >> sep2 >> date.tm_year;
    if (ss.fail() || sep1 != '/' || sep2 != '/') {
        throw runtime_error("Formato de data inválido.");
    }

    date.tm_mon -= 1;
    date.tm_year -= 1900;

    return date;
}